neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py

```diff
@@ -1654,10 +1654,23 @@ def process_network_ports(self, port_info, ovs_restarted):
         self._add_port_tag_info(need_binding_devices)
         if security_disabled_ports:
             added_ports -= set(security_disabled_ports)
-        self.sg_agent.setup_port_filters(added_ports,
-                                         port_info.get('updated', set()))
+
+        # The following call was split into two calls so that added_ports
+        # can be bound quicker.
+        #self.sg_agent.setup_port_filters(added_ports,
+        #                                 port_info.get('updated', set()))
+
+        devices_to_refilter, global_refresh_firewall = (
+            self.sg_agent.setup_port_filters_part_1(added_ports))
+
         failed_devices['added'] |= self._bind_devices(need_binding_devices)
```

Comment:
There was a community patch which did exactly what you suggest: setup_port_filters() was not called until after _bind_devices().   This patch was merged and then later reverted because it opened a security window in which the port would be placed on the network without any security group defined for it.  

Based on DEBUG traces of the neutron-ovs-agent that were collected back in Aug/Sept, when a single port is being added, the refresh_firewall() call  in setup_port_filters() is not done.   The prepare_devices_filter()  is all that is needed.   When a single port is being added + other ports are being updated/removed, then refresh_firewall() is called.   In the collected trace, calling refresh_firewall() caused a delay in the call to _bind_devices() by 10x.  

The call _bind_devices() is only needed on added ports because what we are doing is adding the port to the correct VxLAN (tag) inside of OVS.   This must correspond to the segmentation id of the private network the port is being bound to.  Once this bind operation is done, updates to the port (or to the port's security group) do not affect the binding.   The post is still bound to the same OVS tag / private network.